File: src/app/fdctl/src/commands/configure/frank.rs
use std::path::Path;

use super::*;
use crate::security::*;
use crate::utility::*;

const NAME: &str = "frank";
const CNC_APP_SIZE: u32 = 4032;
const POD_SIZE: u32 = 16384;

pub(super) const STAGE: Stage = Stage {
    name: NAME,
    enabled: None,
    // We can't really verify if a frank workspace has been set up correctly, so if we are
    // running it we just recreate it every time.
    always_recreate: true,
    explain_init_permissions: Some(explain_init_permissions),
    explain_fini_permissions: None,
    init: Some(step),
    fini: None,
    check,
};

#[rustfmt::skip]
fn explain_init_permissions(config: &Config) -> Vec<Option<String>> {
    if config.development.netns.enabled {
        vec![
            check_process_cap(NAME, CAP_SYS_ADMIN, "enter a network namespace"),
        ]
    } else {
        vec![]
    }
}

#[rustfmt::skip]
fn step(config: &mut Config) {
    let bin = &config.binary_dir;
    let name = &config.name;
    let prefix = "firedancer";
    let workspace = format!("{}.wksp", &config.name);
    let interface = &config.tiles.quic.interface;

    if config.development.netns.enabled {
        // Enter network namespace for bind. This is only needed for a check that the interface
        // exists.. we can probably skip that.
        set_network_namespace(&format!("/var/run/netns/{}", interface));
    }

    // Main pod and CNC
    let pod = run!("{bin}/fd_pod_ctl new {workspace} {POD_SIZE}");
    let main_cnc = run!("{bin}/fd_tango_ctl new-cnc {workspace} 0 tic {CNC_APP_SIZE}");
    run!("{bin}/fd_pod_ctl insert {pod} cstr {prefix}.main.cnc {main_cnc}");

    // Pack tiles
    let cnc = run!("{bin}/fd_tango_ctl new-cnc {workspace} 0 tic {CNC_APP_SIZE}");
    let mcache = run!("{bin}/fd_tango_ctl new-mcache {workspace} {} 0 0", config.tiles.pack.max_pending_transactions);
    
    let dcache = run!("{bin}/fd_tango_ctl new-dcache {workspace} 4808 {} 1 1 0", config.tiles.pack.max_pending_transactions);
    let (pack_scratch, cu_est_table) = if Path::new(&format!("{bin}/fd_pack_ctl")).exists() {
        // TODO: Always enable this when merged with pack changes in 1.2
        let pack_scratch = run!("{bin}/fd_pack_ctl new-scratch {workspace} {} {}", config.tiles.pack.solana_labs_bank_thread_count, config.tiles.pack.max_pending_transactions);
        let cu_est_table = run!("{bin}/fd_pack_ctl new-cu-est-tbl {workspace} {} {} {}", config.tiles.pack.compute_unit_estimator_table_size, config.tiles.pack.compute_unit_estimator_ema_history, config.tiles.pack.compute_unit_estimator_ema_default);
        (pack_scratch, cu_est_table)
    } else {
        ("unused".to_string(), "unused".to_string())
    };

    let return_fseq = run!("{bin}/fd_tango_ctl new-fseq {workspace} 0");
    run!("{bin}/fd_pod_ctl \
        insert {pod} cstr {prefix}.pack.cnc {cnc} \
        insert {pod} cstr {prefix}.pack.out-mcache {mcache} \
        insert {pod} cstr {prefix}.pack.out-dcache {dcache} \
        insert {pod} cstr {prefix}.pack.scratch {pack_scratch} \
        insert {pod} cstr {prefix}.pack.cu-est-tbl {cu_est_table} \
        insert {pod} cstr {prefix}.pack.return-fseq {return_fseq} \
        insert {pod} ulong {prefix}.pack.bank-cnt {solana_labs_bank_thread_count} \
        insert {pod} ulong {prefix}.pack.txnq-sz {max_pending_transactions} \
        insert {pod} ulong {prefix}.pack.cu-est-tbl-sz {compute_unit_estimator_table_size} \
        insert {pod} uint {prefix}.pack.cu-limit {solana_labs_bank_thread_compute_units_executed_per_second}",
        solana_labs_bank_thread_count=config.tiles.pack.solana_labs_bank_thread_count,
        max_pending_transactions=config.tiles.pack.max_pending_transactions,
        compute_unit_estimator_table_size=config.tiles.pack.compute_unit_estimator_table_size,
        solana_labs_bank_thread_compute_units_executed_per_second=config.tiles.pack.solana_labs_bank_thread_compute_units_executed_per_second);

    // Dedup tiles
    let cnc = run!("{bin}/fd_tango_ctl new-cnc {workspace} 0 tic {CNC_APP_SIZE}");
    let tcache = run!("{bin}/fd_tango_ctl new-tcache {workspace} {} 0", config.tiles.dedup.signature_cache_size);
    let mcache = run!("{bin}/fd_tango_ctl new-mcache {workspace} {} 0 0", config.tiles.verify.receive_buffer_size); // Same as verify tile depth
    let fseq = run!("{bin}/fd_tango_ctl new-fseq {workspace} 0");
    run!("{bin}/fd_pod_ctl \
        insert {pod} cstr {prefix}.dedup.cnc {cnc} \
        insert {pod} cstr {prefix}.dedup.tcache {tcache} \
        insert {pod} cstr {prefix}.dedup.mcache {mcache} \
        insert {pod} cstr {prefix}.dedup.fseq {fseq}");

    let mut verify_info = vec![];

    // Verify tiles
    for i in 0..config.layout.verify_tile_count {
        let in_mcache = run!("{bin}/fd_tango_ctl new-mcache {workspace} {} 0 0", config.tiles.verify.receive_buffer_size);
        let in_dcache = run!("{bin}/fd_tango_ctl new-dcache {workspace} {mtu} {receive_buffer_size} 1 1 {app_size}",
            mtu=config.tiles.verify.mtu,
            receive_buffer_size=config.tiles.verify.receive_buffer_size,
            app_size=config.tiles.verify.receive_buffer_size * 32);
        let in_fseq = run!("{bin}/fd_tango_ctl new-fseq {workspace} 0");
        run!("{bin}/fd_pod_ctl \
            insert {pod} cstr {prefix}.verifyin.v{i}in.mcache {in_mcache} \
            insert {pod} cstr {prefix}.verifyin.v{i}in.dcache {in_dcache} \
            insert {pod} cstr {prefix}.verifyin.v{i}in.fseq {in_fseq}");

        let cnc = run!("{bin}/fd_tango_ctl new-cnc {workspace} 2 tic {CNC_APP_SIZE}");
        let mcache = run!("{bin}/fd_tango_ctl new-mcache {workspace} {} 0 0", config.tiles.verify.receive_buffer_size);
        let dcache = run!("{bin}/fd_tango_ctl new-dcache {workspace} {mtu} {receive_buffer_size} 1 1 0",
            mtu=config.tiles.verify.mtu,
            receive_buffer_size=config.tiles.verify.receive_buffer_size);
        let fseq = run!("{bin}/fd_tango_ctl new-fseq {workspace} 0");
        run!("{bin}/fd_pod_ctl \
            insert {pod} cstr {prefix}.verify.v{i}.cnc {cnc} \
            insert {pod} cstr {prefix}.verify.v{i}.mcache {mcache} \
            insert {pod} cstr {prefix}.verify.v{i}.dcache {dcache} \
            insert {pod} cstr {prefix}.verify.v{i}.fseq {fseq}");

        verify_info.push((in_mcache, in_dcache, in_fseq));
    }

    // QUIC tiles
    for (i, (verify_mcache, verify_dcache, verify_fseq)) in verify_info.into_iter().enumerate() {
        let cnc = run!("{bin}/fd_tango_ctl new-cnc {workspace} 2 tic {CNC_APP_SIZE}");
        let quic = if Path::new(&format!("{bin}/fd_quic_ctl")).exists() {
            // TODO: Always enable this when merged with QUIC changes in 1.1
            run!("{bin}/fd_quic_ctl new-quic {workspace}")
        } else {
            "unused".to_owned()
        };
        let xsk = run!("{bin}/fd_xdp_ctl new-xsk {workspace} 2048 {} {}", config.tiles.quic.xdp_rx_queue_size, config.tiles.quic.xdp_tx_queue_size);
        run!("{bin}/fd_xdp_ctl bind-xsk {xsk} {name} {interface} {i}");
        let xsk_aio = run!("{bin}/fd_xdp_ctl new-xsk-aio {workspace} {} {}", config.tiles.quic.xdp_tx_queue_size, config.tiles.quic.xdp_aio_depth);

        run!("{bin}/fd_pod_ctl
            insert {pod} cstr {prefix}.quic.quic{i}.cnc {cnc} \
            insert {pod} cstr {prefix}.quic.quic{i}.mcache {mcache} \
            insert {pod} cstr {prefix}.quic.quic{i}.dcache {dcache} \
            insert {pod} cstr {prefix}.quic.quic{i}.fseq {fseq} \
            insert {pod} cstr {prefix}.quic.quic{i}.quic {quic} \
            insert {pod} cstr {prefix}.quic.quic{i}.xsk {xsk} \
            insert {pod} cstr {prefix}.quic.quic{i}.xsk_aio {xsk_aio}",
            mcache=verify_mcache,
            dcache=verify_dcache,
            fseq=verify_fseq);
    }

    let listen_address = super::netns::listen_address(config);
    let src_mac_address = super::netns::src_mac_address(config);

    run!("{bin}/fd_pod_ctl \
        insert {pod} cstr {prefix}.quic_cfg.cert_file {cert_file} \
        insert {pod} cstr {prefix}.quic_cfg.key_file {key_file} \
        insert {pod} cstr {prefix}.quic_cfg.ip_addr {ip_addr} \
        insert {pod} ushort {prefix}.quic_cfg.listen_port {listen_port} \
        insert {pod} cstr {prefix}.quic_cfg.src_mac_addr {src_mac_address} \
        insert {pod} ulong {prefix}.quic_cfg.idle_timeout_ms 1000",
        cert_file=format!("{}/cert.pem", &config.scratch_directory),
        key_file=format!("{}/key.pem", &config.scratch_directory),
        ip_addr=listen_address,
        listen_port=config.tiles.quic.listen_port);

    config.frank.pod = pod.split(':').collect::<Vec<&str>>()[1].parse().unwrap();
    config.frank.main_cnc = main_cnc.split(':').collect::<Vec<&str>>()[1]
            .parse()
            .unwrap();
    config.frank.src_mac_address = src_mac_address;
    config.frank.listen_address = listen_address;

    config.dump_to_bash();
}

fn check(_: &Config) -> CheckResult {
    // Partially configured so the runner tries to perform `undo` every time as well.
    partially_configured!("frank must be reconfigured every launch")
}

File: src/app/fdctl/configure/frank.c
#include "configure.h"

#include "../../../tango/fd_tango.h"
#include "../../../tango/quic/fd_quic.h"
#include "../../../tango/xdp/fd_xsk_aio.h"

#include <linux/capability.h>

#define NAME "frank"

static void
init_perm( security_t *     security,
           config_t * const config ) {
  if( FD_UNLIKELY( config->development.netns.enabled ) )
    check_cap( security, NAME, CAP_SYS_ADMIN, "enter a network namespace" );
}

#define INSERTER(arg, align, footprint, new) do {                                   \
    fd_wksp_t * wksp  = fd_wksp_containing( pod );                                  \
    ulong       gaddr = fd_wksp_alloc( wksp, ( align ), ( footprint ), 1 );         \
    void *      shmem = fd_wksp_laddr( wksp, gaddr );                               \
    (void)shmem;                                                                    \
    if( FD_UNLIKELY( !( new ) ) ) FD_LOG_ERR(( "failed to initialize workspace" )); \
    char name[ PATH_MAX ];                                                          \
    va_list args;                                                                   \
    va_start( args, arg );                                                          \
    vsnprintf( name, PATH_MAX, fmt, args );                                         \
    va_end( args );                                                                 \
    char buffer[ FD_WKSP_CSTR_MAX ];                                                \
    if( FD_UNLIKELY( !fd_wksp_cstr( wksp, gaddr, buffer ) ) )                       \
      FD_LOG_ERR(( "failed to get wksp gaddr for pod entry `%s`", name ));          \
    if( FD_UNLIKELY( !fd_pod_insert_cstr( pod, name, buffer ) ) )                   \
      FD_LOG_ERR(( "failed to insert value into pod for `%s`", name ));             \
  } while( 0 )

#define INSERTER2(arg, align, footprint, new) do {                                              \
    fd_wksp_t * wksp  = fd_wksp_containing( pod );                                              \
    ulong       gaddr = fd_wksp_alloc( wksp, ( align ), ( footprint ), 1 );                     \
    void *      shmem = fd_wksp_laddr( wksp, gaddr );                                           \
    if( FD_UNLIKELY( !( new ) ) ) FD_LOG_ERR(( "failed to initialize workspace" ));             \
    char name1[ PATH_MAX ];                                                                     \
    char name2[ PATH_MAX ];                                                                     \
    va_list args;                                                                               \
    va_start( args, arg );                                                                      \
    vsnprintf( name1, PATH_MAX, fmt1, args );                                                   \
    va_end( args );                                                                             \
    va_start( args, arg );                                                                      \
    vsnprintf( name2, PATH_MAX, fmt2, args );                                                   \
    va_end( args );                                                                             \
    char buffer[ FD_WKSP_CSTR_MAX ];                                                            \
    if( FD_UNLIKELY( !fd_pod_insert_cstr( pod, name1, fd_wksp_cstr( wksp, gaddr, buffer ) ) ) ) \
      FD_LOG_ERR(( "failed to initialize workspace" ));                                         \
    if( FD_UNLIKELY( !fd_pod_insert_cstr( pod, name2, fd_wksp_cstr( wksp, gaddr, buffer ) ) ) ) \
      FD_LOG_ERR(( "failed to initialize workspace" ));                                         \
  } while( 0 )

static void cnc( void * pod, char * fmt, ... ) {
  INSERTER( fmt,
            fd_cnc_align    (                                ),
            fd_cnc_footprint( 4032                           ),
            fd_cnc_new      ( shmem, 4032, 0, fd_tickcount() ) );
}

static void mcache( void * pod, char * fmt, ulong depth, ... ) {
  INSERTER( depth,
            fd_mcache_align    (                    ),
            fd_mcache_footprint( depth, 0           ),
            fd_mcache_new      ( shmem, depth, 0, 0 ) );
}

static void mcache2( void * pod, char * fmt1, char * fmt2, ulong depth, ... ) {
  INSERTER2( depth,
             fd_mcache_align    (                    ),
             fd_mcache_footprint( depth, 0           ),
             fd_mcache_new      ( shmem, depth, 0, 0 ) );
}

static void dcache( void * pod, char * fmt, ulong mtu, ulong depth, ulong app_sz, ... ) {
  ulong data_sz = fd_dcache_req_data_sz( mtu, depth, 1, 1 );
  INSERTER( app_sz,
            fd_dcache_align    (                          ),
            fd_dcache_footprint( data_sz, app_sz          ),
            fd_dcache_new      ( shmem,   data_sz, app_sz ) );
}

static void dcache2( void * pod, char * fmt1, char * fmt2, ulong mtu, ulong depth, ulong app_sz, ... ) {
  ulong data_sz = fd_dcache_req_data_sz( mtu, depth, 1, 1 );
  INSERTER2( app_sz,
             fd_dcache_align    (                          ),
             fd_dcache_footprint( data_sz, app_sz          ),
             fd_dcache_new      ( shmem,   data_sz, app_sz ) );
}

static void fseq( void * pod, char * fmt, ... ) {
  INSERTER( fmt,
            fd_fseq_align    (          ),
            fd_fseq_footprint(          ),
            fd_fseq_new      ( shmem, 0 ) );
}

static void fseq2( void * pod, char * fmt1, char * fmt2, ... ) {
  INSERTER2( fmt2,
             fd_fseq_align    (          ),
             fd_fseq_footprint(          ),
             fd_fseq_new      ( shmem, 0 ) );
}

static void tcache( void * pod, char * fmt, ulong depth, ... ) {
  INSERTER( depth,
            fd_tcache_align    (                 ),
            fd_tcache_footprint( depth, 0        ),
            fd_tcache_new      ( shmem, depth, 0 ) );
}

static void quic( void * pod, char * fmt, fd_quic_limits_t * limits, ... ) {
  INSERTER( limits,
            fd_quic_align    (               ),
            fd_quic_footprint( limits        ),
            fd_quic_new      ( shmem, limits ) );
}

static void xsk( void * pod, char * fmt, ulong frame_sz, ulong rx_depth, ulong tx_depth, ... ) {
  INSERTER( tx_depth,
            fd_xsk_align    (                                                            ),
            fd_xsk_footprint( frame_sz, rx_depth, rx_depth, tx_depth, tx_depth           ),
            fd_xsk_new      ( shmem,    frame_sz, rx_depth, rx_depth, tx_depth, tx_depth ) );
}

static void xsk_aio( void * pod, char * fmt, ulong tx_depth, ulong batch_count, ... ) {
  INSERTER( batch_count,
            fd_xsk_aio_align    (                                 ),
            fd_xsk_aio_footprint( tx_depth, batch_count           ),
            fd_xsk_aio_new      ( shmem,    tx_depth, batch_count ) );
}

static void alloc( void * pod, char * fmt, ulong align, ulong sz, ... ) {
  INSERTER( sz, align, sz, 1 );
}

#define VALUE( type, value ) do {                                 \
    char name[ PATH_MAX ];                                        \
    va_list args;                                                 \
    va_start( args, value );                                      \
    vsnprintf( name, PATH_MAX, fmt, args );                       \
    va_end( args );                                               \
    if( FD_UNLIKELY( !fd_pod_insert_##type( pod, fmt, value ) ) ) \
      FD_LOG_ERR(( "failed to initialize workspace" ));           \
  } while( 0 )

static void ulong1( void * pod, char * fmt, ulong value, ... ) {
  VALUE( ulong, value );
}

static void uint1( void * pod, char * fmt, uint value, ... ) {
  VALUE( uint, value );
}

/* need a dummy argument so we can locate va_args (value would be default
   promoted) */
static void ushort1( void * pod, char * fmt, ushort value, ulong dummy, ... ) { \
    char name[ PATH_MAX ];                                                      \
    va_list args;                                                               \
    va_start( args, dummy );                                                    \
    vsnprintf( name, PATH_MAX, fmt, args );                                     \
    va_end( args );                                                             \
    if( FD_UNLIKELY( !fd_pod_insert_ushort( pod, fmt, value ) ) )               \
      FD_LOG_ERR(( "failed to initialize workspace" ));                         \
}

static void buf( void * pod, char * fmt, void * value, ulong sz, ... ) {
  char name[ PATH_MAX ];
  va_list args;
  va_start( args, sz );
  vsnprintf( name, PATH_MAX, fmt, args );
  va_end( args );
  if( FD_UNLIKELY( !fd_pod_insert_buf( pod, fmt, value, sz ) ) )
    FD_LOG_ERR(( "failed to initialize workspace" ));
}

static void
init( config_t * const config ) {
  /* enter network namespace for bind. this is only needed for a check
     that the interface exists.. we can probably skip that */
  enter_network_namespace( config );

  char workspace[ FD_WKSP_CSTR_MAX ];
  snprintf1( workspace, FD_WKSP_CSTR_MAX, "%s.wksp", config->name );

  fd_wksp_t * wksp  = fd_wksp_attach( workspace );
  ulong       gaddr = fd_wksp_alloc( wksp, fd_pod_align(), fd_pod_footprint( 16384 ), 1 );
  void *      pod   = fd_wksp_laddr( wksp, gaddr );
  if( FD_UNLIKELY( !fd_pod_new( pod, 16384 ) ) ) FD_LOG_ERR(( "failed to initialize workspace" ));

  char pod_str[ FD_WKSP_CSTR_MAX ];
  if( FD_UNLIKELY( !fd_wksp_cstr( wksp, gaddr, pod_str ) ) ) FD_LOG_ERR(( "failed to initialize workspace" ));

  cnc   ( pod, "firedancer.main.cnc" );

  cnc   ( pod, "firedancer.pack.cnc" );
  mcache( pod, "firedancer.pack.out-mcache",           config->tiles.pack.max_pending_transactions );
  dcache( pod, "firedancer.pack.out-dcache",           4808, 6 * config->tiles.pack.max_pending_transactions, 0 );  // 6 should be  `(num_bank_threads+1) * max_pending_transactions + num_bank_threads`
  fseq  ( pod, "firedancer.pack.return-fseq" );
  ulong1( pod, "firedancer.pack.bank-cnt",             config->tiles.pack.solana_labs_bank_thread_count );
  ulong1( pod, "firedancer.pack.txnq-sz",              config->tiles.pack.max_pending_transactions );
  ulong1( pod, "firedancer.pack.cu-limit",             config->tiles.pack.solana_labs_bank_thread_compute_units_executed_per_second );
  ulong1( pod, "firedancer.pack.cu-est-tbl.bin-cnt",   4096 );
  ulong1( pod, "firedancer.pack.cu-est-tbl.footprint", 32 + 32 * 4096 );
  alloc ( pod, "firedancer.pack.cu-est-tbl.memory",    32, 32 + 32 * 4096 );

  cnc   ( pod, "firedancer.dedup.cnc",                 pod, wksp );
  tcache( pod, "firedancer.dedup.tcache",              config->tiles.dedup.signature_cache_size );
  mcache( pod, "firedancer.dedup.mcache",              config->tiles.verify.receive_buffer_size );
  fseq  ( pod, "firedancer.dedup.fseq" );

  for( uint i=0; i<config->layout.verify_tile_count; i++ ) {
    mcache2( pod, "firedancer.verifyin.v%uin.mcache",  "firedancer.quic.quic%u.mcache", config->tiles.verify.receive_buffer_size, i );
    dcache2( pod, "firedancer.verifyin.v%uin.dcache",  "firedancer.quic.quic%u.dcache", config->tiles.verify.mtu, config->tiles.verify.receive_buffer_size, config->tiles.verify.receive_buffer_size * 32, i );
    fseq2  ( pod, "firedancer.verifyin.v%uin.fseq",    "firedancer.quic.quic%u.fseq",   i );

    cnc    ( pod, "firedancer.verify.v%u.cnc",         i );
    mcache ( pod, "firedancer.verify.v%u.mcache",      config->tiles.verify.receive_buffer_size, i );
    dcache ( pod, "firedancer.verify.v%u.dcache",      config->tiles.verify.mtu, config->tiles.verify.receive_buffer_size, 0, i );
    fseq   ( pod, "firedancer.verify.v%u.fseq",        i );
  }

  fd_quic_limits_t limits = {
    .conn_cnt                                      = config->tiles.quic.max_concurrent_connections,
    .handshake_cnt                                 = config->tiles.quic.max_concurrent_handshakes,
    .conn_id_cnt                                   = config->tiles.quic.max_concurrent_connection_ids_per_connection,
    .conn_id_sparsity                              = 0.0,
    .inflight_pkt_cnt                              = config->tiles.quic.max_inflight_quic_packets,
    .tx_buf_sz                                     = config->tiles.quic.tx_buf_size,
    .stream_cnt[ FD_QUIC_STREAM_TYPE_BIDI_CLIENT ] = 0,
    .stream_cnt[ FD_QUIC_STREAM_TYPE_BIDI_SERVER ] = 0,
    .stream_cnt[ FD_QUIC_STREAM_TYPE_UNI_CLIENT  ] = config->tiles.quic.max_concurrent_streams_per_connection,
    .stream_cnt[ FD_QUIC_STREAM_TYPE_UNI_SERVER  ] = 0,
  };

  for( uint i=0; i<config->layout.verify_tile_count; i++) {
    cnc    ( pod, "firedancer.quic.quic%u.cnc",     i );
    quic   ( pod, "firedancer.quic.quic%u.quic",    &limits, i );
    xsk    ( pod, "firedancer.quic.quic%u.xsk",     2048, config->tiles.quic.xdp_rx_queue_size, config->tiles.quic.xdp_tx_queue_size, i );
    xsk_aio( pod, "firedancer.quic.quic%u.xsk_aio", config->tiles.quic.xdp_tx_queue_size, config->tiles.quic.xdp_aio_depth, i );

    char quic_xsk[ FD_WKSP_CSTR_MAX ];
    snprintf1( quic_xsk, FD_WKSP_CSTR_MAX, "firedancer.quic.quic%u.xsk", i );
    char const * quic_xsk_gaddr = fd_pod_query_cstr( pod,  quic_xsk, NULL );
    void *       shmem          = fd_wksp_map      ( quic_xsk_gaddr );
    if( FD_UNLIKELY( !fd_xsk_bind( shmem, config->name, config->tiles.quic.interface, i ) ) )
      FD_LOG_ERR(( "failed to bind xsk for quic tile %u", i ));
    fd_wksp_unmap( shmem );
  }

  uint1  ( pod, "firedancer.quic_cfg.ip_addr",         config->tiles.quic.ip_addr );
  ushort1( pod, "firedancer.quic_cfg.listen_port",     config->tiles.quic.listen_port, 0 );
  buf    ( pod, "firedancer.quic_cfg.src_mac_addr",    config->tiles.quic.mac_addr, 6 );
  ulong1 ( pod, "firedancer.quic_cfg.idle_timeout_ms", 1000 );

  char const * main_cnc = fd_pod_query_cstr( pod, "firedancer.main.cnc", NULL );
  if( FD_UNLIKELY( !main_cnc) )
    FD_LOG_ERR(( "failed to find main cnc in workspace" ));

  dump_vars( config, pod_str, main_cnc );
}

static configure_result_t
check( config_t * const config ) {
  (void)config;
  /* partially configured so the runner tries to perform `undo` every
     time as well */
  PARTIALLY_CONFIGURED( "frank must be reconfigured every launch" );
}

configure_stage_t frank = {
  .name            = NAME,
  /* we can't really verify if a frank workspace has been set up
     correctly, so if we are running it we just recreate it every time */
  .always_recreate = 1,
  .enabled         = NULL,
  .init_perm       = init_perm,
  .fini_perm       = NULL,
  .init            = init,
  .fini            = NULL,
  .check           = check,
};

#undef NAME
